File: python/xfa/04_flow_layout/04_flow_layout.py
# ============================================================================
#  04_flow_layout -- Python (ctypes) port of
#  examples\delphi\xfa\04_flow_layout\04_flow_layout.dpr
#
#  LumasPDF XFA dynamic engine "flavor tour" example 4 of 10 -- FLOW LAYOUT
#  (layout="tb" vertical stacking + layout="lr-tb" left-to-right wrapping).
#  Renders a one-page "Employment Application" with two sibling flowed
#  subforms: TermsPanel (layout="tb", 6 clauses stacked with zero gap) and
#  SkillsPanel (layout="lr-tb", 9 tags wrapping across lines).
#
#    pdfNewPDF -> pdfCreateNewPDFA -> pdfCreateXFAStreamA('template',...) ->
#    pdfCreateXFAStreamA('datasets',...) -> pdfRenderXFAForm -> pdfCloseFile
#
#  Packets are read pre-split from their own .template.xml/.datasets.xml
#  files -- no XML library needed in this driver.
# ============================================================================
import os
import sys
# the wrapper lives at <root>/wrappers/python (source checkout) or beside
# the example tree (shipped package) -- find it without a hard-coded path
_d = os.path.dirname(os.path.abspath(__file__))
for _ in range(6):
    for _c in (os.path.join(_d, 'wrappers', 'python'), _d):
        if os.path.isfile(os.path.join(_c, 'lumaspdf.py')):
            sys.path.insert(0, _c)
            break
    else:
        _d = os.path.dirname(_d)
        continue
    break
import lumaspdf as L

HERE = os.path.dirname(os.path.abspath(__file__))


def render(template_path, datasets_path, out_pdf_path):
    template = open(template_path, "rb").read()
    datasets = open(datasets_path, "rb").read()
    print("template packet bytes:", len(template))
    print("datasets packet bytes:", len(datasets))

    pdf = L.pdfNewPDF()
    if not pdf:
        print("pdfNewPDF FAILED")
        return -100
    try:
        if L.pdfCreateNewPDFA(pdf, out_pdf_path.encode("latin-1")) == 0:
            print("pdfCreateNewPDFA FAILED")
            return -100

        idx = L.pdfCreateXFAStreamA(pdf, b"template", template, len(template))
        print("pdfCreateXFAStreamA(template) -> index", idx)
        if idx < 0:
            print("pdfCreateXFAStreamA(template) FAILED")
            return -100

        idx = L.pdfCreateXFAStreamA(pdf, b"datasets", datasets, len(datasets))
        print("pdfCreateXFAStreamA(datasets) -> index", idx)
        if idx < 0:
            print("pdfCreateXFAStreamA(datasets) FAILED")
            return -100

        rc = L.pdfRenderXFAForm(pdf)
        print("pdfRenderXFAForm ->", rc)
        if rc < 1:
            print("pdfRenderXFAForm FAILED, code", rc)
            return rc

        if L.pdfCloseFile(pdf) == 0:
            print("pdfCloseFile FAILED")
            return -101

        print("OK: wrote", out_pdf_path)
        return rc
    finally:
        L.pdfDeletePDF(pdf)


def main():
    template_path = os.path.join(HERE, "04_flow_layout.template.xml")
    datasets_path = os.path.join(HERE, "04_flow_layout.datasets.xml")
    out_pdf_path = os.path.join(HERE, "04_flow_layout.pdf")

    print("=== 04_flow_layout -> 04_flow_layout.pdf ===")
    rc = render(template_path, datasets_path, out_pdf_path)
    print("RESULT|04_flow_layout=", rc)

    print()
    print("Expected layout geometry (hand-derived, verified against the engine's")
    print("own xfa_layout_dump.exe self-oracle in the Delphi flavor's README):")
    print("  TermsPanel (tb, x=36 y=92 w=540, child h=24, gap=0):")
    print("    Clause1..6 y = 92, 116, 140, 164, 188, 212  (curY += 24 each step)")
    print("  SkillsPanel (lr-tb, x=36 y=270 w=540, child w=110 h=20):")
    print("    line 1 (y=270): Skill1 x=36, Skill2 x=146, Skill3 x=256, Skill4 x=366")
    print("    line 2 (y=290): Skill5 x=36, Skill6 x=146, Skill7 x=256, Skill8 x=366")
    print("    line 3 (y=310): Skill9 x=36")
    print("  3 lines total (4 + 4 + 1 tags), 21 boxes total incl. form1/headers.")


if __name__ == "__main__":
    main()

File: csharp/xfa/04_flow_layout/04_flow_layout.cs
// ============================================================================
//  04_flow_layout -- C# port of examples\delphi\xfa\04_flow_layout
//  XFA "flavor tour" example 4 of 10: flow layout. A one-page "Employment
//  Application" with two sibling flowed subforms inside a layout="position"
//  root -- TermsPanel (layout="tb", 6 clauses stacked vertically, zero gap)
//  and SkillsPanel (layout="lr-tb", 9 tags packed left-to-right, wrapping to
//  a new line when the next tag would overflow the content width).
//
//  Same pipeline as every example in this tour:
//    pdfNewPDF -> pdfCreateNewPDFA -> pdfCreateXFAStreamA('template',...) ->
//    pdfCreateXFAStreamA('datasets',...) -> pdfRenderXFAForm -> pdfCloseFile
// ============================================================================
using System;
using System.IO;
using System.Runtime.InteropServices;
using LumasPdfSdk;

class FlowLayout
{
    static int RenderExample(string exeDir, string baseName, string outPdfPath)
    {
        string templatePath = Path.Combine(exeDir, baseName + ".template.xml");
        string datasetsPath = Path.Combine(exeDir, baseName + ".datasets.xml");

        Console.WriteLine("=== " + baseName + " -> " + Path.GetFileName(outPdfPath) + " ===");
        if (!File.Exists(templatePath)) { Console.WriteLine("FILE-NOT-FOUND: " + templatePath); return -100; }

        byte[] templateBuf = File.ReadAllBytes(templatePath);
        byte[] datasetsBuf = File.Exists(datasetsPath) ? File.ReadAllBytes(datasetsPath) : new byte[0];
        Console.WriteLine("template packet bytes: " + templateBuf.Length);
        Console.WriteLine("datasets packet bytes: " + datasetsBuf.Length);

        IntPtr pdf = LumasPdf.pdfNewPDF();
        if (pdf == IntPtr.Zero) { Console.WriteLine("pdfNewPDF FAILED"); return -100; }
        try
        {
            if (!LumasPdf.pdfCreateNewPDFW(pdf, outPdfPath)) { Console.WriteLine("pdfCreateNewPDFA FAILED"); return -100; }

            GCHandle hTpl = GCHandle.Alloc(templateBuf, GCHandleType.Pinned);
            int idx;
            try { idx = LumasPdf.pdfCreateXFAStreamW(pdf, "template", hTpl.AddrOfPinnedObject(), (uint)templateBuf.Length); }
            finally { hTpl.Free(); }
            Console.WriteLine("pdfCreateXFAStreamA(template) -> index " + idx);
            if (idx < 0) { Console.WriteLine("pdfCreateXFAStreamA(template) FAILED"); return -100; }

            if (datasetsBuf.Length > 0)
            {
                GCHandle hDs = GCHandle.Alloc(datasetsBuf, GCHandleType.Pinned);
                try { idx = LumasPdf.pdfCreateXFAStreamW(pdf, "datasets", hDs.AddrOfPinnedObject(), (uint)datasetsBuf.Length); }
                finally { hDs.Free(); }
                Console.WriteLine("pdfCreateXFAStreamA(datasets) -> index " + idx);
                if (idx < 0) { Console.WriteLine("pdfCreateXFAStreamA(datasets) FAILED"); return -100; }
            }
            else
            {
                Console.WriteLine("(no datasets packet found -- template-only render)");
            }

            int rc = LumasPdf.pdfRenderXFAForm(pdf);
            Console.WriteLine("pdfRenderXFAForm -> " + rc + " (expected: page count >= 1)");
            if (rc < 1) { Console.WriteLine("RENDER-FAILED, code " + rc); return rc; }

            if (!LumasPdf.pdfCloseFile(pdf)) { Console.WriteLine("pdfCloseFile FAILED"); return -101; }
            Console.WriteLine("OK: wrote " + outPdfPath + " (" + rc + " page(s))");
            return rc;
        }
        finally
        {
            LumasPdf.pdfDeletePDF(pdf);
        }
    }

    static void Main()
    {
        try
        {
            string exeDir = AppDomain.CurrentDomain.BaseDirectory;
            int rc = RenderExample(exeDir, "04_flow_layout", Path.Combine(exeDir, "04_flow_layout.pdf"));
            Console.WriteLine();
            Console.WriteLine("RESULT|04_flow_layout=" + rc);
            if (rc >= 1)
            {
                Console.WriteLine("OK: rendered " + rc + " page(s). Hand-computed geometry to confirm:");
                Console.WriteLine("  TermsPanel (tb, x=36 w=540, each child h=24, gap=0):");
                Console.WriteLine("    Clause1..6 y = 92, 116, 140, 164, 188, 212 (curY += 24 each step)");
                Console.WriteLine("  SkillsPanel (lr-tb, x=36 y=270 w=540, each tag w=110 h=20):");
                Console.WriteLine("    Line1 (y=270): Skill1..4 x = 36, 146, 256, 366");
                Console.WriteLine("    Line2 (y=290): Skill5..8 x = 36, 146, 256, 366 (wraps: 476+110=586>540... 5th tag)");
                Console.WriteLine("    Line3 (y=310): Skill9    x = 36");
                Console.WriteLine("  3 lines total (4+4+1 tags), 21 layout boxes overall (verified via xfa_layout_dump.exe oracle).");
            }
            else
            {
                Console.WriteLine("FAILED, see errors above.");
            }
        }
        catch (Exception e)
        {
            Console.WriteLine("EXCEPTION: " + e.GetType().Name + ": " + e.Message);
        }
    }
}
